ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -332,14 +332,21 @@ extern "C" void *__ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[],
             // If there's a reference to a GlobalVariable, copy it over to the
             // new module.
             GlobalVariable *OldGV = cast<GlobalVariable>(Op);
-            GlobalVariable *GV = new GlobalVariable(
-                *JITMod, OldGV->getValueType(), OldGV->isConstant(),
-                OldGV->getLinkage(), (Constant *)nullptr, OldGV->getName(),
-                (GlobalVariable *)nullptr, OldGV->getThreadLocalMode(),
-                OldGV->getType()->getAddressSpace());
-            GV->copyAttributesFrom(&*OldGV);
-            VMap[&*OldGV] = GV;
-            cloned_globals.push_back(OldGV);
+            if (OldGV->isConstant()) {
+              // Global variable is a constant so just copy it into the trace.
+              GlobalVariable *GV = new GlobalVariable(
```

Comment:
It might not be a problem here, but in C (dunno about C++) the evaluation order of paramaters is undefined i.e. "OldGV->isConstant()` might be called before `OldGV->getValueType()`. My sort-of hack for this is that if there's more than one function call in the parameters (and I think there *might* be mutation in one of them), I tend to assign everything to temporary variables before the function call. Possibly more for reference here than anything else.

Reply:
I'm not saying that LLVM is the poster child of code quality, but I've copied this almost verbatim from https://github.com/llvm/llvm-project/blob/d480f968ad8b56d3ee4a6b6df5532d485b0ad01e/llvm/lib/Transforms/Utils/CloneModule.cpp#L63, so I think this should be fine.

I also don't quite understand why it should make any difference if the parameters are evaluated in a different order in this case.

Reply:
If none of the functions does any mutation, it should be safe.

Reply:
As far as I can tell these are all just reads.